### PR TITLE
Handle `SFTPSensorAsync` when file pattern is not passed

### PR DESCRIPTION
**Pull request by bharanidharan14**
- Handled SFTP async operator when file pattern is not passed
- Added `newer_than` attribute corresponding to Sync operators and code changes for the same.
- Added test case for the code changes
closes: #707

### CODE REVIEW

**Comment by bharanidharan14** [comment]
@pankajastro @rajaths010494 @pankajkoti @phanikumv Please review this PR

File: tests/sftp/triggers/test_sftp.py [hunk 1]
```diff
@@ -59,6 +102,30 @@ async def test_sftp_trigger_run_trigger_defer_state(self, mock_get_file_by_patte
         assert task.done() is False
         asyncio.get_event_loop().stop()
 
+    @pytest.mark.asyncio
+    @mock.patch("astronomer.providers.sftp.hooks.sftp.SFTPHookAsync.get_mod_time")
+    async def test_sftp_success_with_newer_then_false(self, mock_mod_time):
+        """
+        Assert that a TriggerEvent with a success status is yielded if a file
+        matching without the pattern
+        """
```
Comment (rajaths010494) [inline tests/sftp/triggers/test_sftp.py:111]: Add appropriate doc string

File: tests/sftp/triggers/test_sftp.py [hunk 2]
```diff
@@ -41,6 +46,44 @@ async def test_sftp_trigger_run_trigger_success_state(self, mock_get_file_by_pat
...
+
+        assert TriggerEvent(expected_event) == actual_event
+
+    @pytest.mark.asyncio
+    @mock.patch("astronomer.providers.sftp.hooks.sftp.SFTPHookAsync.get_mod_time")
+    async def test_sftp_success_with_newer_then(self, mock_mod_time):
+        """
+        Assert that a TriggerEvent with a success status is yielded if a file
+        matching without the pattern
+        """
```
Comment (rajaths010494) [inline tests/sftp/triggers/test_sftp.py:73]: Add appropriate doc string

File: astronomer/providers/sftp/hooks/sftp.py [hunk 3]
```diff
@@ -125,3 +126,20 @@ async def get_file_by_pattern(self, path: str = "", fnmatch_pattern: str = "") -
                 return file
 
         raise AirflowException(f"No files matching file pattern were found at {path} — Deferring")
+
+    async def get_mod_time(self, path: str) -> str:
+        """
+        Returns modification time.
```
[the remark this answers is not in the document]
Comment (bharanidharan14) [reply astronomer/providers/sftp/hooks/sftp.py:132]: File modification time

File: astronomer/providers/sftp/hooks/sftp.py [hunk 4]
```diff
@@ -125,3 +126,20 @@ async def get_file_by_pattern(self, path: str = "", fnmatch_pattern: str = "") -
...
+
+        :param path: full path to the remote file
+        """
+        ssh_conn = await self._get_conn()
+        sftp_client = await ssh_conn.start_sftp_client()
+        try:
+            ftp_mdtm = await sftp_client.stat(path)
+            modified_time = ftp_mdtm.mtime
+            mod_time = datetime.fromtimestamp(modified_time).strftime("%Y%m%d%H%M%S")  # type: ignore[arg-type]
```
[the remark this answers is not in the document]
Comment (bharanidharan14) [reply astronomer/providers/sftp/hooks/sftp.py:141]: `modified_time` is float and `datetime.fromtimestamp()` is expecting int it. It need to be type cast to int

File: astronomer/providers/sftp/triggers/sftp.py [hunk 5]
```diff
@@ -53,13 +58,21 @@ async def run(self) -> AsyncIterator["TriggerEvent"]:  # type: ignore[override]
...
         exc = None
-
+        _newer_than = convert_to_utc(self.newer_than) if self.newer_than else None
         while True:
             try:
-                file_returned_by_hook = await hook.get_file_by_pattern(
-                    path=self.path, fnmatch_pattern=self.file_pattern
-                )
-                yield TriggerEvent({"status": "success", "message": f"Sensed file: {file_returned_by_hook}"})
+                actual_file_to_check = self.path
```
[the remark this answers is not in the document]
Comment (bharanidharan14) [reply astronomer/providers/sftp/triggers/sftp.py:64]: Done

File: astronomer/providers/sftp/hooks/sftp.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bharanidharan14) [reply astronomer/providers/sftp/hooks/sftp.py:132]: Now added proper doc string

File: tests/sftp/hooks/test_sftp.py [hunk 6]
```diff
@@ -156,3 +165,27 @@ async def test_get_file_by_pattern_with_no_match(self, mock_hook_get_conn):
...
+        hook = SFTPHookAsync()
+        mod_time = await hook.get_mod_time("/path/exists/file")
+        expected_value = datetime.datetime.fromtimestamp(1667302566).strftime("%Y%m%d%H%M%S")
+        assert mod_time == expected_value
+
+    @patch("astronomer.providers.sftp.hooks.sftp.SFTPHookAsync._get_conn")
+    @pytest.mark.asyncio
+    async def test_get_mod_time_exception(self, mock_hook_get_conn):
+        """
+        Assert that get_mod_time with mocking the value and exception
```
[the remark this answers is not in the document]
Comment (bharanidharan14) [reply tests/sftp/hooks/test_sftp.py:185]: Fixed

File: tests/sftp/hooks/test_sftp.py [hunk 7]
```diff
@@ -156,3 +165,27 @@ async def test_get_file_by_pattern_with_no_match(self, mock_hook_get_conn):
...
+        """
+        Assert that file attribute and return the modified time of the file
+        """
+        mock_hook_get_conn.return_value.start_sftp_client.return_value = MockSFTPClient()
+        hook = SFTPHookAsync()
+        mod_time = await hook.get_mod_time("/path/exists/file")
+        expected_value = datetime.datetime.fromtimestamp(1667302566).strftime("%Y%m%d%H%M%S")
+        assert mod_time == expected_value
+
+    @patch("astronomer.providers.sftp.hooks.sftp.SFTPHookAsync._get_conn")
```
[the remark this answers is not in the document]
Comment (bharanidharan14) [reply tests/sftp/hooks/test_sftp.py:181]: Fixed

File: tests/sftp/hooks/test_sftp.py [hunk 8]
```diff
@@ -156,3 +165,27 @@ async def test_get_file_by_pattern_with_no_match(self, mock_hook_get_conn):
             await hook.get_file_by_pattern(path="/path/exists/", fnmatch_pattern="file_does_not_exist")
 
         assert str(exc.value) == "No files matching file pattern were found at /path/exists/ — Deferring"
+
+    @patch("astronomer.providers.sftp.hooks.sftp.SFTPHookAsync._get_conn")
```
[the remark this answers is not in the document]
Comment (bharanidharan14) [reply tests/sftp/hooks/test_sftp.py:169]: Fixed

File: tests/sftp/triggers/test_sftp.py (hunk 1 above)
Comment (bharanidharan14) [reply tests/sftp/triggers/test_sftp.py:111]: Fixed

File: tests/sftp/triggers/test_sftp.py (hunk 2 above)
Comment (bharanidharan14) [reply tests/sftp/triggers/test_sftp.py:73]: Fixed

**Comment by bharanidharan14** [comment]
@pankajastro @rajaths010494  Addressed your review comments, please review it

**Review by rajaths010494 — COMMENTED** [review]
Is this added into our master dag to run daily?

**Comment by bharanidharan14** [comment]
> Is this added into our master dag to run daily?

There is a separate ticket for that to cover https://app.zenhub.com/workspaces/airflow-providers-team-61a89ee7fbf048001dc8378d/issues/astronomer/astronomer-providers/662

**Review by rajaths010494 — APPROVED** [review]
LGTM

**Pull request merged**